File: scripts/serve-lan.py
#!/usr/bin/env python3
"""Static file server for LAN preview, with caching disabled.

Same as `python3 -m http.server`, except every response gets
Cache-Control: no-store so editing CSS/JS/HTML is reflected on the next
reload — no hard refresh needed, which matters most from a phone on the
same network, invoked by scripts/serve-lan.js.
"""

import http.server
import os
import socketserver
import sys

PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 3025
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


class NoCacheHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def end_headers(self):
        self.send_header("Cache-Control", "no-store, no-cache, must-revalidate")
        self.send_header("Pragma", "no-cache")
        self.send_header("Expires", "0")
        super().end_headers()


class ReusableTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    """Threaded on purpose: TCPServer answers one request at a time, so the page's assets
    queue up behind each other. With no-store on every response that queue is paid at each
    reload, and the stylesheet can land after the browser has given up waiting and painted
    the page unstyled."""

    allow_reuse_address = True
    daemon_threads = True


if __name__ == "__main__":
    with ReusableTCPServer(("0.0.0.0", PORT), NoCacheHandler) as httpd:
        httpd.serve_forever()

File: scripts/serve-lan.js
#!/usr/bin/env node
//# SERVE LAN
//+ Wraps scripts/serve-lan.py (a no-cache static server) used to browse the demo from another
//+ device on the same network (a phone, say): prints the LAN URL for the pre-baked demo page before
//+ starting the server, so there's no need to remember/type the path by hand. Plain
//+ `python3 -m http.server` sends no Cache-Control header, so a phone browser keeps serving a stale
//+ CSS/JS/HTML from cache until a hard refresh — serve-lan.py disables caching instead.

"use strict";

const path = require("path");
const os = require("os");
const { spawn } = require("child_process");

function getLanIp() {
    const interfaces = os.networkInterfaces();
    for (const name of Object.keys(interfaces)) {
        for (const iface of interfaces[name] || []) {
            if (iface.family === "IPv4" && !iface.internal) return iface.address;
        }
    }
    return null;
}

const PORT = 3025;
const DEMO_PATH = "/site/site.html";
const lanIp = getLanIp();

console.log(`[serve:lan] http://127.0.0.1:${PORT}${DEMO_PATH}`);
if (lanIp) console.log(`[serve:lan] http://${lanIp}:${PORT}${DEMO_PATH}`);

spawn("python3", [path.join(__dirname, "serve-lan.py"), String(PORT)], { stdio: "inherit" });
